src/MarshalMiddlewareTrait.php

```diff
@@ -35,52 +38,55 @@
-     * @return callable
+     * @return ServerMiddlewareInterface
      * @throws Exception\InvalidMiddlewareException
      */
-    private function prepareMiddleware($middleware, ContainerInterface $container = null, $forError = false)
-    {
-        if (is_callable($middleware)) {
+    private function prepareMiddleware(
+        $middleware,
+        Router\RouterInterface $router,
+        ResponseInterface $responsePrototype,
+        ContainerInterface $container = null
+    ) {
+        if ($middleware === Application::ROUTING_MIDDLEWARE) {
+            return new Middleware\RouteMiddleware($router, $responsePrototype);
+        }
+
+        if ($middleware === Application::DISPATCH_MIDDLEWARE) {
+            return new Middleware\DispatchMiddleware();
+        }
+
+        if ($middleware instanceof ServerMiddlewareInterface) {
             return $middleware;
         }
 
-        if ($middleware === Application::ROUTING_MIDDLEWARE) {
-            return [$this, 'routeMiddleware'];
+        if ($this->isCallableInteropMiddleware($middleware)) {
+            return new CallableInteropMiddlewareWrapper($middleware);
         }
 
-        if ($middleware === Application::DISPATCH_MIDDLEWARE) {
-            return [$this, 'dispatchMiddleware'];
+        if (is_callable($middleware)) {
+            return new CallableMiddlewareWrapper($middleware, $responsePrototype);
         }
 
         if (is_array($middleware)) {
-            return $this->marshalMiddlewarePipe($middleware, $container, $forError);
+            return $this->marshalMiddlewarePipe($middleware, $router, $responsePrototype, $container);
         }
 
         if (is_string($middleware) && $container && $container->has($middleware)) {
-            $method = $forError ? 'marshalLazyErrorMiddlewareService' : 'marshalLazyMiddlewareService';
-            return $this->{$method}($middleware, $container);
+            return new Middleware\LazyLoadingMiddleware($container, $responsePrototype, $middleware);
         }
 
-        $callable = $middleware;
         if (is_string($middleware)) {
-            $callable = $this->marshalInvokableMiddleware($middleware);
+            return $this->marshalInvokableMiddleware($middleware, $responsePrototype);
         }
 
-        if (! is_callable($callable)) {
-            throw new Exception\InvalidMiddlewareException(
-                sprintf(
-                    'Unable to resolve middleware "%s" to a callable',
-                    (is_object($middleware)
-                    ? get_class($middleware) . "[Object]"
-                    : gettype($middleware) . '[Scalar]')
-                )
-            );
-        }
-
-        return $callable;
+        throw new Exception\InvalidMiddlewareException(sprintf(
+            'Unable to resolve middleware "%s" to a callable or %s',
+            (is_object($middleware) ? get_class($middleware) . "[Object]" : gettype($middleware) . '[Scalar]'),
```

Comment:
Useless parentheses around the whole expression, and please change `"` to `'` around `[Object]`.